src/devices/bus/a2bus/a2frob.cpp

```diff
@@ -0,0 +1,430 @@
+  to be a device instead of a driver.
+
+  The slot interface for the a2600 has been removed for simplicity as well as
+  the PAL a2600 variant.
+
+  Note: Once the Frob memory is loaded, you can reset the 2600 with Numpad Minus
+
+***************************************************************************/
+
+
+#include "emu.h"
+
+#include "a2bus.h"
+#include "bus/vcs/rom.h"
+#include "bus/vcs/vcs_slot.h"
+#include "bus/vcs_ctrl/ctrl.h"
+#include "cpu/m6502/m6507.h"
+#include "emupal.h"
+#include "screen.h"
+#include "softlist_dev.h"
+#include "sound/tiaintf.h"
+#include "speaker.h"
+#include "../../mame/atari/tia.h"
+
+#include "machine/mos6530.h"
+
+//#define VERBOSE (LOG_GENERAL)
+#include "logmacro.h"
+
+#include "a2frob.h"
+
+
+namespace {
+
+static const uint16_t supported_screen_heights[4] = { 262, 312, 328, 342 };
+
+class a2600_frob_base_device :
+	public device_t,
+	public device_a2bus_card_interface
+{
+
+public:
+	a2600_frob_base_device(const machine_config &mconfig, const char *tag, device_t *owner, u32 clock);
+
+	INPUT_CHANGED_MEMBER(reset_a2600);
+
+protected:
+	a2600_frob_base_device(const machine_config &mconfig, device_type type, const char *tag, device_t *owner, u32 clock) :
+	device_t(mconfig, type, tag, owner, clock),
+		device_a2bus_card_interface(mconfig, *this),
+		m_riot_ram(*this, "riot_ram"),
+		m_tia(*this, "tia_video"),
+		m_maincpu(*this, "maincpu"),
+		m_riot(*this, "riot"),
+		m_joy1(*this, "joyport1"),
+		m_joy2(*this, "joyport2"),
+		m_screen(*this, "tia_screen"),
+		m_xtal(3.579575_MHz_XTAL)
+	{ }
+
+	virtual void device_start() override ATTR_COLD;
+
+	void a2600_mem(address_map &map) ATTR_COLD;
+
+	virtual tiny_rom_entry const *device_rom_region() const override ATTR_COLD;
+	virtual void device_add_mconfig(machine_config &config) override ATTR_COLD;
+	virtual ioport_constructor device_input_ports() const override ATTR_COLD;
+
+	virtual u8 read_c0nx(u8 offset) override;
+	virtual void write_c0nx(u8 offset, u8 data) override;
+
+	virtual u8 read_cnxx(u8 offset) override
+	{ return !m_frob_apple_control ? m_frob_mem[(offset & 0xff)|(m_frob_page << 8)] : 0; }
+
+	virtual void write_cnxx(u8 offset, u8 data) override
+	{ if (!m_frob_apple_control) m_frob_mem[(offset & 0xff)|(m_frob_page << 8)] = data; }
+
+	virtual bool take_c800() override { return false; }
+
+	u8 m_frob_mem[0x1000] = { 0 };  // 4k frob memory
+
+	u8 read_frob_mem(offs_t offset) { return !m_frob_apple_control ? 0 : m_frob_mem[offset & 0xfff]; }
+
+	void write_frob_mem(offs_t offset, u8 data) { m_frob_mem[offset & 0xfff] = data; }
+
+public:
+	void a2600_write_fff0(offs_t offset, u8 data)
+	{
+		m_byte_for_apple = data;
+		m_byte_waiting_flag_apple = 1;
+	}
+
+	u8 a2600_read_fff2(offs_t offset)
+	{
+		if (!machine().side_effects_disabled()) m_byte_waiting_flag_vcs = 0;
+		return m_byte_for_vcs;
+	}
+
+	u8 a2600_read_fff1(offs_t offset)
+	{
+		return (m_byte_waiting_flag_apple ? 0 : 1) << 7 | (m_byte_waiting_flag_vcs << 6);
+	}
```

Comment:
Why have you made this stuff `public`?

Also, you need to use `synchronize` barriers to get data from one CPU to the other to avoid things being seen in the wrong order.  The card CPU should see the flag indicating that the mailbox for the host CPU is full as soon as it writes it, but the host CPU should not see the flag or the updated value until your on the other side of a sync barrier.  The same applies for communication in the opposite direction.

Reply:
synchronize() doesn't work the opposite direction*, some drivers do trickery with machine().scheduler().perfect_quantum(x) to make sure the 'opposite' side comms answer timing is correct.

*:
scheduler runs CPUs in serial, CPU A, then CPU B.
let's say CPU A: 1ms.
scheduler runs CPU B.
At 0.5ms point, CPU B wants to write to a latch that shared with CPU A. CPU A however, is already in the future (from CPU B perspective)

Reply:
And to solve that properly (since you don’t know in advance whether a CPU is going to attempt to communicate across scheduling domains), you need an efficient way to roll back and retry a smaller timeslice until you get an “acceptable” delay.  But since the CPUs themselves don’t know who’s ahead at any given time, putting both directions through `synchronize` at least means that whichever CPU is ahead will try to end its timeslice early when possible.

Reply:
Is it possible for CPU B to be ahead of CPU A? And I don't mean slight deviations with icount < 0.

eg. you have 3 CPUs in a driver. Then the scheduler runs the CPUs in this order A,B.C.A,B,C,A,B,C, etc. (or A,B,C,B,C,B,C if CPU B does a synchronize()). It doesn't eenie meenie miney mo them. CPU B is never ahead of CPU A. CPU C is never ahead of CPU A or B.
The order is the same as the machine config order. It's why for arcade drivers with a 1-way soundlatch, it's important to list the audiocpu after maincpu.